File: pingpong/src/pingpong.erl
% Compile under Linux and macOS with "erlc -W  -o pingpong.beam  pingpong.erl"
% Run with "erl -noshell -run pingpong start -s init stop -rounds 1000"
%
% Asynchronous Message passing is used for sending the "ball" between the ping and the
% pong Erlang processes. Messages from the ping process to the pong process include
% the process id over which the pong process sends the "ball" back.
%
% In Erlang, a program terminates when the main process terminates. Here, the main/pong
% processes wait for the ping process to terminate.

-module(pingpong).
-export([start/0, ping/1, pong/1]).

start()->
    {ok, [Rounds | _]} = init:get_argument(rounds),
    {R, _} = string:to_integer(Rounds),
    Pong = spawn(pingpong, pong, [self()]),
    Ping = spawn(pingpong, ping, [self()]),
    Ping ! {Pong, R},
    receive
        done ->
            skip
    end.
    
ping(Main) ->
    receive
        {Pong, Bounces} ->
            if Bounces > 0 ->
                % io:fwrite("Ping: " ++ integer_to_list(Bounces) ++ "\n"),
                Pong ! {self(), Bounces - 1},
                ping(Main); 
            true ->
                Pong ! done,
                Main ! done
            end      
    end.

pong(Main) ->
    receive
        {Ping, Bounces} ->
            if Bounces >= 0 ->
                % io:fwrite("Pong: " ++ integer_to_list(Bounces) ++ "\n"),
                Ping ! {self(), Bounces - 1}, 
                pong(Main);
            true ->
                receive 
                    done ->
                        skip
                end
            end
    end.

File: pingpong/src/pingpong.hs
{- Compile under Linux and macOS with
"ghc -threaded -o pingpong --make -O -rtsopts pingpong.hs";
run with "./pingpong 10000"

Asynchronous channels with capacity of 1 (MVar) are used for sending the "ball"
between the ping and the pong threads. The threads do not "know" about each
other, they communicate only through three channels: pingChannel and
pongChannel to integer type and doneChannel of unit type.

In Haskell, a program terminates when the main thread terminates. Here, the
main thread and pong thread wait for the notification from the ping threads.
-}
import Control.Concurrent
import System.Environment

ping :: (MVar Int) -> (MVar Int) -> (MVar Bool) -> IO()
ping pingChan pongChan doneChan = do        
    bounces <- takeMVar pongChan
    if bounces > 0
        then do
            putMVar pingChan (bounces - 1)
            ping pingChan pongChan doneChan
        else do
            putMVar doneChan True
            putMVar doneChan True
        
pong :: (MVar Int) -> (MVar Int) -> (MVar Bool) -> IO()
pong pingChan pongChan doneChan = do
    bounces <- takeMVar pingChan
    if bounces >= 0
        then do
            putMVar pongChan (bounces - 1)
            pong pingChan pongChan doneChan
        else do
            _ <- takeMVar doneChan
            return()

main = do
    args <- getArgs
    let bounces = read (args!!0)
    pingChan <- newEmptyMVar
    pongChan <- newEmptyMVar
    doneChan <- newEmptyMVar
    forkIO $ ping pingChan pongChan doneChan
    forkIO $ pong pingChan pongChan doneChan
    putMVar pingChan bounces
    _ <- takeMVar doneChan
    return ()